### PR TITLE
Completely disable all relevant build steps of extensions (except FeatureBuildItem production) when their "enabled" configuration switch is set to false

### PR DESCRIPTION
**Pull request by yrodiere**
In particular, when such an extension is disabled:

* ~~Don't display the feature in logs on startup~~ => changed following comments, see conversation below.
* Don't enable SSL
* Don't add reflective classes in native mode

EDIT: Also, regardless of whether the extension is enabled or not, always display the feature in logs on startup.

~~Creating this PR as draft, because it's based on #26965 , which must be merged first.~~ => Done

I separated this PR from #26965 because #26965 doesn't alter the behavior of any extension, while this PR does (but for the better, IMO).

### CODE REVIEW

File: extensions/micrometer/deployment/src/main/java/io/quarkus/micrometer/deployment/MicrometerConfigUnremovableProcessor.java
Comment (ebullient) [inline extensions/micrometer/deployment/src/main/java/io/quarkus/micrometer/deployment/MicrometerConfigUnremovableProcessor.java:15]: I think this build step can remain in the main Processor.. it was not originally tagged as "only if micrometer was enabled" .. but classes/beans that need the micrometer config are not going to be loaded if micrometer is disabled...

Comment (yrodiere) [reply extensions/micrometer/deployment/src/main/java/io/quarkus/micrometer/deployment/MicrometerConfigUnremovableProcessor.java:15]: That's what I thought initially, but then I looked at the git history, and this was added specifically to prevent an NPE when micrometer is disabled: https://github.com/quarkusio/quarkus/pull/25169 . Apparently `RestClientMetricsListener` is still around in that case and it will trigger loading of the `MicrometerConfig` class.

So, we should probably run this build step when Micrometer is disabled?

Comment (ebullient) [reply extensions/micrometer/deployment/src/main/java/io/quarkus/micrometer/deployment/MicrometerConfigUnremovableProcessor.java:15]: why... oh... because it's a stupid service loader thing?

Some people love service loader, but I'm not a fan, for reasons like this. I can't just turn it off.. the config has to be preserved so the automagic thing discovered in the jar knows whether or not to do anything. :sigh: I wonder if there is a better way to do that?

**Comment by gsmet** [comment]
> Don't display the feature in logs on startup

Can't say I'm excited about this particular one. For me, it's important to know that the feature has actually be detected by Quarkus even if not active.

**Comment by yrodiere** [comment]
> Can't say I'm excited about this particular one. For me, it's important to know that the feature has actually be detected by Quarkus even if not active.

We'll have to make up our minds then, because other extensions (Micrometer, ...) hide the feature when they are disabled:

https://github.com/quarkusio/quarkus/blob/8816386b87e145ade6ef55aa0d7572f20c6eec21/extensions/micrometer/deployment/src/main/java/io/quarkus/micrometer/deployment/MicrometerProcessor.java#L86-L89

For context, here's the log message we're talking about:

> 2022-07-27 15:31:25,238 INFO  [io.quarkus] (Quarkus Main Thread) Installed features: [agroal, cdi, hibernate-orm, hibernate-orm-panache, hibernate-search-elasticsearch, jdbc-postgresql, narayana-jta, resteasy-reactive, resteasy-reactive-jackson, smallrye-context-propagation, vertx]

We can argue whether it's clear that "installed" doesn't mean "enabled". Personally I find the distinction too confusing and I would hide the extensions when they are disabled, but in the end I don't really care either way, as long as we're consistent across Quarkus.

FWIW, we also use `FeatureBuildItem` to forbid having two (enabled?) extensions with the same name:

https://github.com/quarkusio/quarkus/blob/c09cf508e87f2f5826ae46d2a4341146a1e5b914/core/deployment/src/main/java/io/quarkus/deployment/steps/MainClassBuildStep.java#L257-L264

Not sure what to think about that. We probably wouldn't want to have two extensions with the same name even if one is disabled?

**Comment by gsmet** [comment]
Well I don't know about Micrometer but the purpose of this message has always been to inform about the features detected by Quarkus.
Enabled or not, that's a matter of config.

We can decide to change that. But... my personal opinion is that it's not a good idea. Knowing that an extension has actually been detected is an important piece of information and why we have this line in the log.

> FWIW, we also use FeatureBuildItem to forbid having two (enabled?) extensions with the same name

I think this is probably something that predates capabilities. But it's still a good thing in my opinion as having 2 features with the same name would be confusing.

**Comment by yrodiere** [comment]
Okay, well, I'll let you argue with @ebullient on the expected behavior :)

I have no horse in this game, I'll just make everything consistent once you agree.

FWIW, the oidc client extensions also hide the feature when they are disabled.

I'll just suggest a third solution: we could display `(disabled)` next to the name of an extension in the log message.

**Comment by ebullient** [comment]
> Well I don't know about Micrometer but the purpose of this message has always been to inform about the features detected by Quarkus. Enabled or not, that's a matter of config.
> 
> We can decide to change that. But... my personal opinion is that it's not a good idea. Knowing that an extension has actually been detected is an important piece of information and why we have this line in the log.

I've always only registered the micrometer feature if active. I guess the meaning was confusing for me, too. (installed to me was relevant at runtime, where I try to strip out as much as possible at build time if it has been disabled)


> 
> > FWIW, we also use FeatureBuildItem to forbid having two (enabled?) extensions with the same name
> 
> I think this is probably something that predates capabilities. But it's still a good thing in my opinion as having 2 features with the same name would be confusing.

**Comment by gsmet** [comment]
So just to be clear: my point is that I don't want us to change the meaning of a very visible feature (pun intended) in a somehow unrelated pull request.
I would rather align Micrometer with the rest of Quarkus and the original intent of this log line and IF we want to change this behavior for something better, have a discussion on the list about this specific thing.

It might look like a detail but it's not: we don't log many things with Quarkus and the meaning of what we log and how it is supposed to be consumed by users is important.

**Comment by ebullient** [comment]
> So just to be clear: my point is that I don't want us to change the meaning of a very visible feature (pun intended) in a somehow unrelated pull request. I would rather align Micrometer with the rest of Quarkus and the original intent of this log line and IF we want to change this behavior for something better, have a discussion on the list about this specific thing.
> 
> It might look like a detail but it's not: we don't log many things with Quarkus and the meaning of what we log and how it is supposed to be consumed by users is important.

Happy to align micrometer -- just pointing out that I missed the meaning when creating the extension in the first place (Liberty also logged installed features... but only those that are active.. so my base was a little different)

**Comment by ebullient** [comment]
> Especially since I expect some bikeshedding about the output.
🤣

File: extensions/micrometer/deployment/src/main/java/io/quarkus/micrometer/deployment/MicrometerConfigUnremovableProcessor.java
Comment (ebullient) [inline extensions/micrometer/deployment/src/main/java/io/quarkus/micrometer/deployment/MicrometerConfigUnremovableProcessor.java:16]: If we rename this class... "AlwaysOnProcessor?!?!" .. the feature thing could go here.

**Comment by yrodiere** [comment]
I applied the changes, now all extensions produce a FeatureBuildItem regardless of whether they are enabled or not.

I'll chase after extension owners later to get their approval.

**Review by radcortez — APPROVED** [review]
Approve the OTel changes.

**Review by ebullient — APPROVED** [review]
Approve micrometer changes

File: extensions/micrometer/deployment/src/main/java/io/quarkus/micrometer/deployment/MicrometerConfigUnremovableProcessor.java
Comment (ebullient) [reply extensions/micrometer/deployment/src/main/java/io/quarkus/micrometer/deployment/MicrometerConfigUnremovableProcessor.java]: Aside, I have removed the 🤬 service loader here (for other reasons): https://github.com/quarkusio/quarkus/pull/26978.

**Comment by yrodiere** [comment]
Hey all,

So, one of the main changes here is that features will be displayed in logs on startup, even for extensions disabled through configuration properties. I'm talking about this log message:

```
2022-07-27 15:31:25,238 INFO [io.quarkus] (Quarkus Main Thread) Installed features: [agroal, cdi, hibernate-orm, hibernate-orm-panache, hibernate-search-elasticsearch, jdbc-postgresql, narayana-jta, resteasy-reactive, resteasy-reactive-jackson, smallrye-context-propagation, vertx]
```

The new behavior is compliant with what the team had in mind when this log message was introduced and what we currently expect from extensions (don't believe me, believe @gsmet :) ).

I already got approval for the Micrometer and OTel/Jaeger extensions (thanks!).

@sberyozkin/@pedroigor, could you please confirm this isn't a problem for the `keycloak-authorization` extension? For the various `oidc` extensions?

@FroMage could you please confirm this isn't a problem for the `security-webauthn` extension?

**Comment by yrodiere** [comment]
> @sberyozkin could you please confirm this isn't a problem for the various `oidc` extensions?

Hey @sberyozkin, bump :) Any opinion on this change for the `oidc` extensions? If you don't particularly care either way, I think we'll go ahead, since other extension owners agree with the change.

**Comment by pedroigor** [comment]
@yrodiere I'm really happy to see this change happening. It would help a lot Keycloak and open the doors to lot of improvements. One example is https://github.com/quarkiverse/quarkus-vault/issues/12. IIRC, this PR should help with that.

As per the security extensions, I would leave it to @sberyozkin. But I don't see an issue for `keycloak-authorization`.

**Comment by pedroigor** [comment]
@yrodiere I guess the changes here are not yet allowing disabling build steps conditionally, right? For instance, Keycloak is basically a Quarkus extension and we would like to conditionally disable build steps depending on the user configuration when running re-augmentation.

**Comment by yrodiere** [comment]
> @yrodiere I'm really happy to see this change happening. It would help a lot Keycloak and open the doors to lot of improvements. For instance, [quarkiverse/quarkus-vault#12 (comment)](https://github.com/quarkiverse/quarkus-vault/issues/12#issuecomment-991140157). AFAIK, this PR should help with that.

I'm not sure it will, since this PR only disables a few build steps when the (existing) `*.enabled` configuration properties are set to false. The `@BuildSteps` annotation has already been merged in a previous PR, so you can probably already experiment with it. It's mostly just syntactic sugar, though; see #26965 / #26980 .

As to disabling extensions at runtime, which I think [was mentioned in the issue you linked](https://github.com/quarkiverse/quarkus-vault/issues/12#issuecomment-992305787), that's still not possible, unfortunately. But maybe `@BuildSteps` alone is enough?

**Comment by yrodiere** [comment]
> @yrodiere I guess the changes here are not yet allowing disabling build steps conditionally, right?

This PR doesn't change anything regarding that, no.

> For instance, Keycloak is basically a Quarkus extension and we would like to conditionally disable build steps depending on the user configuration when running re-augmentation.

That was already possible, I think? Just use `@BuildStep(onlyIf = MyCondition.class)` on that step and have the relevant configuration injected in `MyCondition`. See https://quarkus.io/guides/writing-extensions#conditional-step-inclusion

Recently I added `@BuildSteps` (with a trailing `s`) that can be applied at the type level, which is more convenient, but is just syntactic sugar.

**Comment by pedroigor** [comment]
> Recently I added `@BuildSteps` (with a trailing `s`) that can be applied at the type level, which is more convenient, but is just syntactic sugar.

Yeah, using a condition to a `BuildStep` works and we are using that. But I did not know about the top-level class `BuildSteps`.

Can I expect that when an extension is using `BuildSteps` the `ExtensionLoader` won't even try to resolve methods from the build class using reflection?

**Comment by pedroigor** [comment]
Btw, also related to what we are looking for https://github.com/quarkusio/quarkus/issues/22122.

**Comment by pedroigor** [comment]
@yrodiere Sure, thanks for your input. This change is still valuable and we should be looking at how adapt to it.

**Comment by gsmet** [comment]
@yrodiere I think we should pursue with this before we end up with conflicts.

**Comment by yrodiere** [comment]
> > @sberyozkin could you please confirm this isn't a problem for the various `oidc` extensions?
> 
> Hey @sberyozkin, bump :) Any opinion on this change for the `oidc` extensions? If you don't particularly care either way, I think we'll go ahead, since other extension owners agree with the change.

No answer, so I'll assume it's good to go.

Thanks everyone! Merging

**Comment by yrodiere** [comment]
... I'll let CI run first though :-)

**Review by brunobat — APPROVED** [review]

**Pull request merged**
